### ISSUE TITLE
may randomly fail

### ISSUE THREAD

**Issue by torsten-deriv** [Grinnz/Mojolicious-Plugin-SetUserGroup#2]
I have seen the following message generated by the module:

```
Can't switch to group "nogroup": No child processes
```

Now, that error (`ECHILD`) does not make any sense in `setgid` context. So, I inspected the code:

``` perl
setgid($gid);
return _error($app, qq{Can't switch to group "$group": $!}) if $!;
```

That's definitive the wrong way to do error checking. `errno` (or `$!`) is meant to be inspected if you need more detailed information about the error. The fact that an error has occurred is conveyed by the return code of the syscall. In fact, `errno` is not changed if a syscall was successful. It is only set on failure.

So, the correct way would be something along these lines:

``` perl
return _error($app, qq{Can't switch to group "$group": $!})
    if setgid($gid) < 0;
return _error($app, qq{Can't set supplemental groups "@groups": $!})
    if setgroups(@gids) < 0;
return _error($app, qq{Can't switch to user "$user": $!})
    if setuid($uid) < 0;
```

**Comment by Grinnz** [comment]
Thanks for the report. setgroups() from Unix::Groups appears to return false on failure rather than -1. I will also add better error checking for the getpwnam related syscalls, this seems to require setting $! to 0 before each call.

**Comment by Grinnz** [comment]
This has been released as 0.004.
